providers/base/bin/camera_test.py

```diff
@@ -282,89 +370,204 @@ def _setup(self, sink=None):
+        """
+        Setup the gstreamer pipeline to capture an image.
+        This pipeline consists of the following elements:
+        - v4l2src: Capture video from a V4L2 device
+        - caps: Filter the video stream to the desired resolution
+        - bayer2rgb (only for bayer format): Convert the bayer format to RGB
+        - jpegenc: Encode the video stream to a JPEG image
+        - filesink: Save the JPEG image to a file
+        """
+
+        pipeline = self.Gst.Pipeline.new("pipeline")
+
+        # Add source
+        source = self.Gst.ElementFactory.make("v4l2src", "video-source")
+        source.set_property("device", self.device)
+        source.set_property("num-buffers", 10)
+        pipeline.add(source)
+
+        # Add caps
+        caps = self.Gst.ElementFactory.make("capsfilter", "caps")
+        if pixelformat == "RG10":
+            caps.set_property(
+                "caps",
+                self.Gst.Caps.from_string(
+                    "video/x-bayer, format=gbrg, width={},height={}".format(
+                        width, height
+                    )
+                ),
+            )
+            # Add bayer2rgb
+            bayer2rgb = self.Gst.ElementFactory.make("bayer2rgb", "bayer2rgb")
+
+            pipeline.add(caps)
+            pipeline.add(bayer2rgb)
+        else:
+            caps.set_property(
+                "caps",
+                self.Gst.Caps.from_string(
+                    "video/x-raw,width={},height={}".format(width, height)
+                ),
+            )
+            pipeline.add(caps)
+
+        # Add encoder
+        encoder = self.Gst.ElementFactory.make("jpegenc", "encoder")
+        pipeline.add(encoder)
+
+        # Add sink
+        sink = self.Gst.ElementFactory.make("filesink", "sink")
+        sink.set_property("location", filename)
+        pipeline.add(sink)
+
+        # Link elements
+        source.link(caps)
+        # If the pixelformat is RG10, we need to link the bayer2rgb element
+        if pixelformat == "RG10":
+            caps.link(bayer2rgb)
+            bayer2rgb.link(encoder)
+        else:
+            caps.link(encoder)
+        encoder.link(sink)
+
+        # Connect the bus to the message handler
+        bus = pipeline.get_bus()
+        bus.add_signal_watch()
+        bus.connect("message", self._on_gst_message)
+
+        # Start the pipeline
+        self.pipeline = pipeline
+        self.pipeline.set_state(self.Gst.State.PLAYING)
+
+        # Add a timeout of 90 seconds to capture the image
+        self.timeout = self.GLib.timeout_add_seconds(90, self._on_timeout)
+
+        # Start the main loop
+        self.main_loop = self.GLib.MainLoop()
+        try:
+            self.main_loop.run()
+        except self.GLib.Error:
+            self.main_loop.quit()
+            self.pipeline.set_state(self.Gst.State.NULL)
+
+        # Remove the timeout
+        if self.timeout:
+            self.GLib.source_remove(self.timeout)
+        return 0
```

Comment:
Please raise a `TimeoutError` and remove that `return 0`

Reply:
(seems to be that the parent here doesn't handle the timeout and this is most likely a bug!

Reply:
So maybe this is a bit confusing, but here, I am just removing the function watching for the timeout (`self.timeout`) so It is cleared for the next iteration.
According to [gtk documentation](https://docs.gtk.org/glib/main-loop.html#memory-management-of-sources) that's the way to go:
```
The first, and preferred, option is to store the source ID returned by functions such as g_timeout_add() or g_source_attach(), and explicitly remove that source from the main context using g_source_remove() when the owning object is finalized.
```
Maybe I can make it more clear on the description

Reply:
Ok this wasn't clear to me, should have read this better. What happens when the timeout expires? Does it raise an error somehow? I don't understand how. That is what I was tring to suggest here, what does that return 0 mean? That return is the only one in this function, so what happens when this fails? I dont undestand